Add optimization for shift by CNS_INT

adiaaida opened the pull request:
This change adds support for shifting by a GT_CNS_INT without going
through a helper. If the shiftOp is a GT_CNS_INT we do several
transformations based on the shift amount:

If the shift amount is 0, the shift is a nop, so we just put together the
hi and lo ops as a GT_LONG.

If the shift amount is < 32, we generate a shl/shld pattern, a shr/shrd
pattern or a sar/shrd pattern, depending on the oper. The first operand of
the shrd/shld is a GT_LONG, which we crack in codegen, using it
essentially as two int operands, rather than creating a tri op GenTree
node (essentially so that we can have 3 operands, instead of the normal
two).

If the shift amount is 32, it differs between shifting left and shifting
right. For GT_LSH, we move the loOp into the hiResult and set the loResult
to 0. For GT_RSZ, we move the hiOp into the loResult, and set the hiResult
to 0. For GT_RSH, we move the hiOp into the loResult, and set the hiResult
to a 31 bit signed shift of the hiOp to sign extend.

If the shift amount is less than 64, but larger than 32: for GT_LSH, the
hiResult is a shift of the loOp by shift amount - 32 (the move from lo into hi is
the 32 bit shift). We set the loResult to 0. For GT_RSH and GT_RSZ, the
loResult is a right shift (signed for GT_RSH) of the hiOp by shift amount
- 32. The hiResult is 0 for GT_RSZ, and a 31 bit signed shift of hiOp1 for
GT_RSH.

If the shift amount is >= 64, we set both hiResult and loResult to 0 for
GT_LSH and GT_RSZ, and do a sign extend shift to set hiResult and loResult
to the sign of the original hiOp for GT_RSH.

adiaaida commented:
@BruceForstall @pgavlin @CarolEidt @sivarv PTAL
/cc: @dotnet/jit-contrib

BruceForstall commented:
There are a lot of different, interesting cases here. Do we have existing test cases that cover all of them? Did you see the expected codegen in all cases?

This would be an interesting change to get asm diffs for, to see how many you get, and verify they're what you expect.

adiaaida commented:
I didn't do asm diffs, but I will. Thanks for the suggestion.

For some background, I took a look at the code patterns generated by codegenlegacy for these cases and tried to match them as closely as possible. All of the different cases can be found in legacy.

BruceForstall commented on src/jit/decomposelongs.cpp:
```diff
@@ -920,15 +925,24 @@ GenTree* DecomposeLongs::DecomposeShift(LIR::Use& use)
+
     genTreeOps oper = tree->OperGet();
+    genTreeOps shiftOper = oldShiftOp->OperGet();
+
```
The name "shiftOper" is confusing to me. How about "shiftByOper"?

BruceForstall replied:
even more confusing is that below I see it referred to as "shiftWidthUse" which, IMO, should be shiftByUse.

---
In reply to: [78653812](https://github.com/dotnet/coreclr/pull/7169#discussion_r78653812) [](ancestors = 78653812)

BruceForstall replied:
and shiftWidthOp. How about shiftCountOper, shiftCountUse, oldShiftCountOp, shiftCountOp, etc?

---
In reply to: [78654147](https://github.com/dotnet/coreclr/pull/7169#discussion_r78654147) [](ancestors = 78654147,78653812)

BruceForstall commented on src/jit/decomposelongs.cpp:
```diff
@@ -920,15 +925,24 @@ GenTree* DecomposeLongs::DecomposeShift(LIR::Use& use)
     GenTree* tree   = use.Def();
     GenTree* gtLong = tree->gtGetOp1();
+    GenTree* oldShiftOp = tree->gtGetOp2();
+
```
Do we actually need this, since we have shiftWidthOp, below, which looks like the same thing?

adiaaida replied:
Shiftwidth op is a lclvar. Every time I tried to get the CNS_INT that the lcl var stored, I got compile errors. Also, in the case of shift being a constant, I don't think we actually need the store to lcl var, so maybe that should actually be moved down to the else case.

BruceForstall commented on src/jit/decomposelongs.cpp:
```diff
@@ -920,15 +925,24 @@ GenTree* DecomposeLongs::DecomposeShift(LIR::Use& use)
+        count = oldShiftOp->gtIntCon.gtIconVal;
+    }
 
     assert((oper == GT_LSH) || (oper == GT_RSH) || (oper == GT_RSZ));
```
can this all be within the shiftOper 'if', below? Did you put it here before shiftWidthOp got Remove()'d? But it already gets used after the remove.

adiaaida replied:
Done.

BruceForstall commented on src/jit/decomposelongs.cpp:
```diff
@@ -964,20 +975,252 @@ GenTree* DecomposeLongs::DecomposeShift(LIR::Use& use)
+    if (shiftOper == GT_CNS_INT)
+    {
+        GenTree* loCopy  = m_compiler->gtNewLclvNode(loOp1LclNum, TYP_INT);
 
```
It looks like this isn't always used. You should only create the new lclvar node when it is actually required.

adiaaida replied:
done

BruceForstall commented on src/jit/decomposelongs.cpp:
```diff
@@ -964,20 +975,252 @@ GenTree* DecomposeLongs::DecomposeShift(LIR::Use& use)
 
-    GenTreeArgList* argList = m_compiler->gtNewArgList(loOp1, hiOp1, shiftWidthOp);
+    // If we are shifting by a constant int, we do not want to save to a helper, instead, we decompose.
+    if (shiftOper == GT_CNS_INT)
```
Can't all the helper choosing code above be pushed into the "else" case that actually creates the helper call?

adiaaida replied:
done

BruceForstall commented on src/jit/decomposelongs.cpp:
```diff
@@ -964,20 +975,252 @@ GenTree* DecomposeLongs::DecomposeShift(LIR::Use& use)
+                    // reg1 = lo
+                    // shl lo, shift
+                    // shld hi, reg1, shift
+
```
Don't we want to produce:
```
shld hi, lo, shift
shl lo, shift
```
?

adiaaida replied:
With how LIR works, or at least, how I understand it to work, we can't use lo in 2 places, so we have to store a copy of it anyway. Since we store a copy of it, the instructions can be generated in either order. In FinalizeDecomposition, we assert that the hiResult comes after the loResult, so lo must be put in LIR first, then hi, so we're going to generate the instructions in the opposite order (shl first, then shld).

BruceForstall replied:
I think we need to fix FinalizeDecomposition() to take an insertAfter argument, and not assert anything about the relative ordering of loResult and hiResult. Then I think you could simply insert the high nodes before the low nodes in linear order. You do need an additional lclvar node, so you do need loCopy (it's not actually a copy of the value, but an additional lclvar reference). You need to call `m_compiler->lvaIncRefCnts(loCopy);` to account for the additional lclvar reference in the LSH_HI tree.


---
In reply to: [78656652](https://github.com/dotnet/coreclr/pull/7169#discussion_r78656652) [](ancestors = 78656652)

adiaaida replied:
By adding an insertAfter argument, you mean changing the signature to:

`GenTree* DecomposeLongs::FinalizeDecomposition(LIR::Use& use, GenTree* loResult, GenTree* hiResult, GenTree* insertAfter)`

And most calls to it would be changed to:

`FinalizeDecomposition(use, loResult, hiResult, hiResult)`

Except in the LSH case, where we'd call:

`FinalizeDecomposition(use, loResult, hiResult, loResult)`

BruceForstall replied:
Yes. Maybe to be clear the param should be named insertResultAfter?

adiaaida replied:
Sounds good.

adiaaida replied:
Done.

BruceForstall commented on src/jit/decomposelongs.cpp:
```diff
@@ -964,20 +975,252 @@ GenTree* DecomposeLongs::DecomposeShift(LIR::Use& use)
+                    Range().InsertAfter(tree, loOp1, shiftByLo, loResult);
+                    Range().InsertAfter(loResult, loCopy, hiOp1, hiOp);
+                    Range().InsertAfter(hiOp, shiftByHi, hiResult);
+                }
```
As a general point of style, one thing you _could_ do (certainly not necessary) is convert all the inserts to InsertBefore(), in which case you could probably always use `tree` as the insert point, instead of needing to change the insert point as you add additional insertions. This is maybe more useful in the >=32 case below, where the insertions are not all together.

BruceForstall commented on src/jit/decomposelongs.cpp:
```diff
@@ -964,20 +975,252 @@ GenTree* DecomposeLongs::DecomposeShift(LIR::Use& use)
+                            GenTree* shiftBy = m_compiler->gtNewIconNode(count - 32, TYP_INT);
+                            hiResult         = m_compiler->gtNewOperNode(oper, TYP_INT, loOp1, shiftBy);
+                            Range().InsertAfter(loResult, shiftBy, loOp1, hiResult);
+                        }
```
Was it intentional that you put the shift count in execution order before loOp1 (a naive tree walk of hiResult would visit loOp1 first)? It probably doesn't matter, since shiftBy will be contained.

adiaaida replied:
No, I meant to put them in the order they're used. Good catch.

BruceForstall commented on src/jit/lowerxarch.cpp:
```diff
@@ -1030,6 +1034,13 @@ void Lowering::TreeNodeInfoInitShiftRotate(GenTree* tree)
 
+#ifdef _TARGET_X86_
+    if (source->OperGet() == GT_LONG)
+    {
```
It would be useful to add a comment here to describe why this is a special case.

adiaaida replied:
Done.

BruceForstall commented on src/jit/codegenxarch.cpp:
```diff
@@ -4715,8 +4725,24 @@ instruction CodeGen::genGetInsForOper(genTreeOps oper, var_types type)
+#else
+    assert(tree->OperIsShiftOrRotate() || tree->OperGet() == GT_LSH_HI || tree->OperGet() == GT_RSH_LO);
+#endif
+
```
Given the number of `#ifdefs` in that you've added, and the lack of commonality with the new code and the old, how about adding instead a new genCodeForShiftLong() function that just handles the two new special HI opcodes?

adiaaida replied:
Good point. Done.

BruceForstall commented on src/jit/codegenxarch.cpp:
```diff
@@ -4728,6 +4754,47 @@ void CodeGen::genCodeForShift(GenTreePtr tree)
+            {
+                inst_RV_RV(INS_mov, tree->gtRegNum, regLo, targetType);
+            }
+        }
```
Might be useful to tail merge this and the above by setting a regResult in each case and then handling the 'mov' just once based on regResult.

adiaaida replied:
Done.

BruceForstall commented:
Generally looking good! Just a few things I'd like to see.

briansull commented on src/jit/codegenxarch.cpp:
```diff
@@ -4715,8 +4725,24 @@ instruction CodeGen::genGetInsForOper(genTreeOps oper, var_types type)
     assert(tree->OperIsShiftOrRotate());
-    assert(!tree->gtOp.gtOp1->isContained());
+#else
+    assert(tree->OperIsShiftOrRotate() || tree->OperGet() == GT_LSH_HI || tree->OperGet() == GT_RSH_LO);
```
Why can't those two new opcodes be added to OperIsShiftOrRotate so that you dont have to special case them here.

adiaaida replied:
I didn't want to trigger genCodeForShiftRMW

adiaaida commented:
I think I captured all of the changes @BruceForstall suggested, including:

1) Move genCodeForShiftLong into its own function.
2) Replace InsertAfters with InsertBefores
3) Modified FinalizeDecompose to have an insertResultAfter parameter and remove the assert that hi follows lo.
4) Fixed the lva ref counts for the copies.
5) Put in some paranoia in lower for this case to mark the gt_long operands as delayRegFree (@sivarv, I may have gone overboard here. Can you take a look at that piece?)
6) With the above change, I was able to move the move in codegen to above the shift, where it should be (and is for the normal shift case).

And probably some other things that I can't recall off the top of my head at this point.

I also did look at some of the asm diffs (though I had some issues with using jit-diff, so I just manually inspected various functions that exercised the various code paths), and they look how I expect them to look.

PTAL

adiaaida commented:
@dotnet/jit-contrib PTAL

sivarv commented on src/jit/codegenlinear.h:
```diff
@@ -146,6 +146,10 @@ void genSetRegToIcon(regNumber reg, ssize_t val, var_types type = TYP_INT, insFl
 
+#if defined(_TARGET_X86_) && !defined(LEGACY_BACKEND)
+void genCodeForShiftLong(GenTreePtr tree);
+#endif
```
(Minor) codegenLinear.h is a header file that is included only in RyuJIT backend.   For that reason we don't need !defined(LEGACY_BACKEND) in #ifdef above.

sivarv commented on src/jit/decomposelongs.cpp:
```diff
@@ -307,17 +307,16 @@ GenTree* DecomposeLongs::DecomposeNode(GenTree* tree)
 //    The next node to process.
 //
-GenTree* DecomposeLongs::FinalizeDecomposition(LIR::Use& use, GenTree* loResult, GenTree* hiResult)
+GenTree* DecomposeLongs::FinalizeDecomposition(LIR::Use& use, GenTree* loResult, GenTree* hiResult, GenTree* insertResultAfter)
```
>, GenTree* insertResultAfter) [](start = 98, length = 29)

(Minor) Function header needs to be updated with the addition of this new arg.

sivarv commented on src/jit/decomposelongs.cpp:
```diff
@@ -920,64 +924,319 @@ GenTree* DecomposeLongs::DecomposeShift(LIR::Use& use)
 
-    // TODO-X86-CQ: If the shift operand is a GT_CNS_INT, we should pipe the instructions through to codegen
-    // and generate the shift instructions ourselves there, rather than replacing it with a helper call.
+    // If we are shifting by a constant int, we do not want to save to a helper, instead, we decompose.
```
>save [](start = 63, length = 4)

(Minor) should comment read as ".. do not want to use a helper" instead of ".. do not want to save to a helper"?

sivarv commented on src/jit/decomposelongs.cpp:
```diff
@@ -920,64 +924,319 @@ GenTree* DecomposeLongs::DecomposeShift(LIR::Use& use)
 
     LIR::Use loOp1Use(Range(), &gtLong->gtOp.gtOp1, gtLong);
-    loOp1Use.ReplaceWithLclVar(m_compiler, m_blockWeight);
+    unsigned loOp1LclNum = loOp1Use.ReplaceWithLclVar(m_compiler, m_blockWeight);
```
Don't we need to check whether lop1 is already an LclVar before replacing with another LclVar?

Same comment applies to line 938

sivarv commented on src/jit/decomposelongs.cpp:
```diff
@@ -920,64 +924,319 @@ GenTree* DecomposeLongs::DecomposeShift(LIR::Use& use)
-    LIR::Use shiftWidthUse(Range(), &tree->gtOp.gtOp2, tree);
-    shiftWidthUse.ReplaceWithLclVar(m_compiler, m_blockWeight);
+    LIR::Use shiftByUse(Range(), &tree->gtOp.gtOp2, tree);
+    shiftByUse.ReplaceWithLclVar(m_compiler, m_blockWeight);
```
We are doing this replacement by local even before checking whether shiftByOper is an int constant.
If it is an int constant, do we need to replace it with an LclVar?

sivarv commented on src/jit/decomposelongs.cpp:
```diff
@@ -920,64 +924,319 @@ GenTree* DecomposeLongs::DecomposeShift(LIR::Use& use)
+                        m_compiler->lvaIncRefCnts(hiCopy);
+                    }
+
+                    else
```
(Minor) unnecessary blank line between preceding "{" and else

sivarv commented on src/jit/lowerxarch.cpp:
```diff
@@ -1030,6 +1034,23 @@ void Lowering::TreeNodeInfoInitShiftRotate(GenTree* tree)
+#ifdef _TARGET_X86_
+    // The first operand of a GT_LSH_HI and GT_RSH_LO oper is a GT_LONG so that
+    // we can have a three operand form. Increment the srcCount.
+    if (source->OperGet() == GT_LONG)
```
>   if (source->OperGet() == GT_LONG) [](start = 1, length = 36)

I would suggest that this condition be

if (tree->OperGet() == GT_LSH_HI || tree->OperGet() == GT_RSH_LO)
{
   assert(source->OperGet() == GT_LONG);
    .....
}

sivarv commented on src/jit/codegenxarch.cpp:
```diff
@@ -4766,6 +4780,66 @@ void CodeGen::genCodeForShift(GenTreePtr tree)
+    if (regResult != tree->gtRegNum)
+    {
+        inst_RV_RV(INS_mov, tree->gtRegNum, regResult, targetType);
+    }
```
Given this logic we can conclude that

if oper=GT_LSH_HI
 then tree->gtRegNum being same as regHi is beneficial as it avoids a move.  This also means that OperandHi doesn't need to be marked as isDelayFree in this case.

If oper = GT_RSH_LO
 then tree->gtRegNum being same as regLo is beneficial as it avoids a move.  This also means that OperandLo doesn't need to be marked as isDelayFree in this case.

adiaaida replied:
Ok. I'll put in that kind of logic in lowerxarch.cpp. Is it only operandHi/operandLo that need to be marked as isDelayFree or does the GT_LONG parent also need to be marked?

sivarv replied:
The parent (GT_LONG) needs to marked as containing delayFree sources (i.e. hasDelayFreeSrc=true)  even if one of its source is marked as isDelayFree.

---
In reply to: [79074794](https://github.com/dotnet/coreclr/pull/7169#discussion_r79074794) [](ancestors = 79074794)

sivarv commented on src/jit/codegenxarch.cpp:
```diff
@@ -4766,6 +4780,66 @@ void CodeGen::genCodeForShift(GenTreePtr tree)
+
+    if (oper == GT_LSH_HI)
+    {
+        inst_RV_RV_IV(ins, EA_4BYTE, tree->gtRegNum, regLo, count);
```
>EA_4BYTE [](start = 27, length = 8)

Why hardcode type as EA_4BYTE, shouldn't this be emitTypeSize(targetType) which is supposed to be EA_4BYTE?

This comment also applies to line 4836 as well

sivarv commented on src/jit/lowerxarch.cpp:
```diff
@@ -1030,6 +1034,23 @@ void Lowering::TreeNodeInfoInitShiftRotate(GenTree* tree)
+
+        source->gtLsraInfo.hasDelayFreeSrc = true;
+        info->hasDelayFreeSrc = true;
+    }
```
HasDelayFreeSrc setting looks correct to me.
Only which of the lo or hi sources gets marked as isDelayFree could be optimized.

adiaaida commented on src/jit/lowerxarch.cpp:
```diff
@@ -1030,6 +1034,23 @@ void Lowering::TreeNodeInfoInitShiftRotate(GenTree* tree)
+    if (source->OperGet() == GT_LONG)
+    {
+        info->srcCount++;
+        source->gtLsraInfo.isDelayFree = true;
```
@sivarv I was actually wondering about if this is necessary. Not only was I marking both hi and lo as isDelayFree, but I've marked the GT_LONG node as isDelayFree. I just want to make sure this is necessary, or remove it if it isn't.

sivarv replied:
I missed that you are marking source (which is GT_LONG)as isDelayFree, which is not required since it will be contained.

What is required here is

Mark one of sourceLo or sourceHi as isDelayFree depending on oper is LSH_HI or RSH_LO.
Mark source and tree as hasDelayFreeSrc=true

---
In reply to: [79076154](https://github.com/dotnet/coreclr/pull/7169#discussion_r79076154) [](ancestors = 79076154)

sivarv commented on src/jit/codegenxarch.cpp:
```diff
@@ -4766,6 +4780,66 @@ void CodeGen::genCodeForShift(GenTreePtr tree)
+
+    GenTreePtr shiftBy = tree->gtGetOp2();
+
+    assert(shiftBy->IsCnsIntOrI());
```
>assert(shiftBy->IsCnsIntOrI()); [](start = 4, length = 31)

Since ShiftBy amount is also supposed to be contained, the assert should be

assert(shiftBy->IsContainedIntOrImmed());

sivarv commented on src/jit/codegenxarch.cpp:
```diff
@@ -4766,6 +4780,66 @@ void CodeGen::genCodeForShift(GenTreePtr tree)
+
+    assert(shiftBy->IsCnsIntOrI());
+
+    int count = (int)shiftBy->AsIntConCommon()->IconValue();
```
>int count [](start = 4, length = 9)

(Minor) inst_RV_RV_IV() takes a an unsigned shift count. Declaring it as unsigned will make the intent clear.
(note sure whether compiler would have given a warning about implicit conversion from int to unsigned)

adiaaida replied:
That was copied from genCodeForShift. Should I update that function too?

sivarv commented on src/jit/gtlist.h:
```diff
@@ -173,6 +173,8 @@ GTNODE(SUB_LO     , "-Lo"          ,0,GTK_BINOP)
 GTNODE(DIV_HI     , "/Hi"          ,0,GTK_BINOP)
 GTNODE(MOD_HI     , "%Hi"          ,0,GTK_BINOP)
+GTNODE(LSH_HI     , "<<Hi"         ,0,GTK_BINOP)
+GTNODE(RSH_LO     , ">>Lo"         ,0,GTK_BINOP)
```
These shift hi/lo opers are different from the rest that these take a long type operand as op1.  It would be a good idea to add a couple of comment lines explaining their purpose and where they are used.

adiaaida commented:
@sivarv I believe I have addressed all of the issues you pointed out. PTAL.

sivarv commented on src/jit/gtlist.h:
```diff
@@ -173,12 +173,23 @@ GTNODE(SUB_LO     , "-Lo"          ,0,GTK_BINOP)
+// the high bits of the lo operand into the high operand as it shifts left. RSH_LO
+// will shift the lo bits of the high operand into the lo operand). LSH_HI
+// represents the high operation of a 64-bit left shift by a constant int, and
+// RSH_LO represents the lo operation of a 64-bit right shift by a contant int.
```
>contant [](start = 67, length = 7)

Typo "constant"

sivarv commented:
:shipit:

sivarv commented on src/jit/lowerxarch.cpp:
```diff
@@ -1030,6 +1034,31 @@ void Lowering::TreeNodeInfoInitShiftRotate(GenTree* tree)
+    {
+        assert(source->OperGet() == GT_LONG);
+
+        info->srcCount++;
```
Something happened to remember signing off:

For codgen we don't need sourceHi to be always in reg in case of GT_LSH_HI (because it could be moved from memory to targetReg if sourceHi is a contained mem-op).  Similarly for GT_RSH_LO, sourceLo could be marked as contained memory-op.  Even if not a memory-op, we could mark it as reg-optional.

It is ok not to implement contained mem-op/reg-optional in the first implementation.  But please capture as a TODO-CQ comment and an issue, so that we can revisit this in future.

adiaaida replied:
Where should the TODO-CQ comment go? In lower or in codegen where we call genCodeForShiftRMW for other shifts?

sivarv replied:
You could put TODO-CQ comment in codegenxarch.cpp just before the new function that was added.    This TODO-CQ applies only to GT_RSH_LO and GT_LSH_HI.

CarolEidt commented on src/jit/codegenxarch.cpp:
```diff
@@ -4766,6 +4780,66 @@ void CodeGen::genCodeForShift(GenTreePtr tree)
+    assert(oper == GT_LSH_HI || oper == GT_RSH_LO);
+
+    GenTree* operand = tree->gtOp.gtOp1;
+    assert(!operand->gtOp.gtOp1->isContained());
```
It would probably be prudent to assert that `operand` is a `GT_LONG`

CarolEidt commented:
@adiaaida Bruce asked about testing. Do you know if all these cases are actually hit by existing tests? It wasn't totally clear. If not, perhaps you could open an issue to verify or add test cases.

CarolEidt commented:
:shipit:

adiaaida commented:
@CarolEidt it appears we do not hit the following cases:

Anything shifted by 0 or >= 64.
RSH by 32.

I'll try to create some test cases for these. I'm assuming they'll have to be IL test.

adiaaida commented:
Opened #7216 to track adding the tests.

Opened #7217 to track contained mem-op/reg optional shifts.
